src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+      return ns.setDefault(js, JsValue(js.wrapBytes(kj::heapArray<kj::byte>(data))));
+    }, [&](Value exception) {
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+      return false;
+    });
+  };
+}
+
+Module::EvaluateCallback Module::newJsonModuleHandler(kj::Array<const char> data) {
+  return [data = kj::mv(data)](Lock& js,
+                               const Url& specifier,
+                               const ModuleNamespace& ns,
+                               const CompilationObserver& observer) -> bool {
+    return js.tryCatch([&] {
+      auto metrics = observer.onJsonCompilationStart(js.v8Isolate, data.size());
+      return ns.setDefault(js, JsValue(js.parseJson(data).getHandle(js)));
+    }, [&](Value exception) {
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+      return false;
+    });
+  };
+}
+
+Module::EvaluateCallback Module::newWasmModuleHandler(kj::Array<kj::byte> data) {
+  struct Cache final {
+    kj::MutexGuarded<kj::Maybe<v8::CompiledWasmModule>> mutex {};
+  };
+  return [data = kj::mv(data), cache = kj::heap<Cache>()](
+      Lock& js,
+      const Url& specifier,
+      const ModuleNamespace& ns,
+      const CompilationObserver& observer) mutable -> bool {
+    return js.tryCatch([&]() -> bool {
+      js.setAllowEval(true);
+      KJ_DEFER(js.setAllowEval(false));
+
+      // Allow Wasm compilation to spawn a background thread for tier-up, i.e. recompiling
+      // Wasm with optimizations in the background. Otherwise Wasm startup is way too slow.
+      // Until tier-up finishes, requests will be handled using Liftoff-generated code, which
+      // compiles fast but runs slower.
+      AllowV8BackgroundThreadsScope scope;
+
+      {
+        // See if we can use a cached compiled module to speed things up.
+        auto lock = cache->mutex.lockShared();
+        KJ_IF_SOME(compiled, *lock) {
+          auto metrics = observer.onWasmCompilationFromCacheStart(js.v8Isolate);
+          auto result = JsValue(check(
+            v8::WasmModuleObject::FromCompiledModule(js.v8Isolate, compiled)));
+          return ns.setDefault(js, result);
+        }
+      }
+
+      auto module = jsg::compileWasmModule(js, data, observer);
+      auto lock = cache->mutex.lockExclusive();
+      *lock = module->GetCompiledModule();
+      auto result = JsValue(module);
+      return ns.setDefault(js, result);
+    }, [&](Value exception) {
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+      return false;
+    });
+  };
+}
+
+Function<void()> Module::compileEvalFunction(
+    Lock& js,
+    kj::StringPtr code,
+    kj::StringPtr name,
+    kj::Maybe<JsObject> compileExtensions,
+    const CompilationObserver& observer) {
+  auto metrics = observer.onScriptCompilationStart(js.v8Isolate, name);
+  v8::ScriptOrigin origin(js.str(name));
+  v8::ScriptCompiler::Source source(js.str(code), origin);
+  auto fn = ([&] {
+    KJ_IF_SOME(ext, compileExtensions) {
```

Comment:
Off-topic: I wish kj::Maybe had something like `unwrap_or` that would allow us to avoid having two check() calls here. Probably worth adding in the future.